crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +503,372 @@ where
+                tracing::info!(
+                    "Wrapper transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg,
+                );
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::InvalidTx));
+                // Make sure to clean the write logs for the next transaction
+                self.state.write_log_mut().drop_tx();
+            }
+            Err(msg) => {
+                // This branch represents an error that affects the entire
+                // batch
+                tracing::info!(
+                    "Transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg
+                );
+
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::WasmRuntimeError));
+
+                self.handle_batch_error(
+                    response,
+                    msg,
+                    tx_result,
+                    tx_data,
+                    &mut tx_logs,
+                );
+            }
+        }
+    }
+
+    // Evaluate the results of all the transactions of the batch. Commit or drop
+    // the storage changes, update stats and event, manage replay protection.
+    fn handle_inner_tx_results(
+        &mut self,
+        response: &mut shim::response::FinalizeBlock,
+        tx_result: namada::tx::data::TxResult<protocol::Error>,
+        tx_data: TxData,
+        tx_logs: &mut TxLogs,
+    ) {
+        // Track the need to commit the batch hash for replay protection. Hash
+        // must be written if at least one of the txs in the batch requires so
+        let mut commit_batch_hash = false;
+        // Track if any of the inner txs failed or was rejected
+        let mut is_any_tx_invalid = false;
+        let mut temp_log = TempTxLogs::default();
+
+        temp_log.check_inner_results(
+            &tx_result,
+            tx_data.tx_header,
+            tx_data.tx_index,
+            tx_data.height,
+            &mut commit_batch_hash,
+            &mut is_any_tx_invalid,
+        );
+
+        if tx_data.is_atomic_batch && is_any_tx_invalid {
+            // Atomic batches need custom handling when even a single tx fails,
+            // since we need to drop everything
+            let unrun_txs = tx_data.commitments_len
+                - tx_result.batch_results.0.len() as u64;
+            temp_log.stats.set_failing_atomic_batch(unrun_txs);
+            temp_log.commit_stats_only(tx_logs);
+            self.state.write_log_mut().drop_batch();
+            tx_logs.tx_event.extend(Code(ResultCode::WasmRuntimeError));
```

Comment:
Why this error code?

Reply:
We've always used this error except for wrapper errors

Reply:
Is there not a more specific error returned from one of the inner transactions inside the batch which failed?

Reply:
Ah yes the specific error can be found in the `Event` we emit under the `"batch"` attribute: here we serialize the entire `TxResult` object which carries information about all the inner txs we ran including possible errors or the vps that rejected them

Reply:
Hmm OK, I still think it would be good to have an error code which actually specifies the semantics in this case, e.g. `ResultCode::AtLeastOneInnerTxFailed` or something - but this can be handled later if it's a pain to do right now.

Reply:
Sounds good, I've opened [this](https://github.com/anoma/namada/issues/3199) issue